src/main/go/org/jenkinsci/plugins/durabletask/heartbeat-launcher.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+var mainLogger *log.Logger
+var hbLogger *log.Logger
+var launchLogger *log.Logger
+var scriptLogger *log.Logger
+
+func checkIfErr(process string, err error) bool {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%v: %v\n", strings.ToUpper(process), err.Error())
+		return true
+	}
+	return false
+}
+
+func checkScriptErr(err error) bool {
+	return loggerIfErr(scriptLogger, err)
+}
+
+func loggerIfErr(logger *log.Logger, err error) bool {
+	if err != nil {
+		logger.Println(err.Error())
+		return true
+	}
+	return false
+}
+
+// Catch a termination signal to allow for a graceful exit (i.e. no zombies)
+func signalCatcher(sigChan chan os.Signal) {
+	for sig := range sigChan {
+		mainLogger.Printf("(sig catcher) caught: %v\n", sig)
+	}
+}
+
+func launcher(wg *sync.WaitGroup, exitChan chan bool, cookieName string, cookieVal string,
+	interpreter string, scriptPath string, resultPath string, outputPath string) {
+
+	defer wg.Done()
+	defer signalFinished(exitChan)
+
+	var scriptCmd *exec.Cmd
+	if interpreter != "" {
+		scriptCmd = exec.Command(interpreter, "-xe", scriptPath)
+	} else {
+		scriptCmd = exec.Command(scriptPath)
+	}
+	cookieString := fmt.Sprintf("%v=%v", cookieName, cookieVal)
+	scriptCmd.Env = append(os.Environ(),
+		cookieString)
+
+	if outputPath != "" {
+		// capturing output
+		outputFile, err := os.Create(outputPath)
+		if checkScriptErr(err) {
+			exitLauncher(-2, resultPath)
+			return
+		}
+		defer outputFile.Close()
+		scriptCmd.Stdout = outputFile
+	} else {
+		scriptCmd.Stdout = scriptLogger.Writer()
+	}
+	if outputPath != "" {
+		// capturing output
+		scriptCmd.Stderr = scriptLogger.Writer()
+	} else {
+		// Note: pointing to os.Stdout will not capture all err logs
+		scriptCmd.Stderr = scriptCmd.Stdout
+	}
+	// Create new session
+	scriptCmd.SysProcAttr = &unix.SysProcAttr{Setsid: true}
+	for i := 0; i < len(scriptCmd.Args); i++ {
+		launchLogger.Printf("args %v: %v\n", i, scriptCmd.Args[i])
```

Comment:
Do the old scripts do this? If not seems better to avoid printing args in case of secrets, although they may be masked depending on exactly what plugins are being used.

Ah ok, so after reading on, it looks like these loggers just act like `/dev/null` unless a debug flag is set?

Reply:
Yes, that is correct, it will only output (to a file) when the debug flag is given.